### PR TITLE
Add required server option

### LINKED ISSUE #1939: Add a "Required" flag for a server
This is a request to add a "Required" flag to a server and appropriate logic, so that if this server stops responding, say due to a billing or login issue, sabnzbd does not blindly proceed to empty a configured block account.

It should instead stall until the "Required" server starts responding again.

### PR DESCRIPTION
**Pull request by puzzledsab**
Fixes #1939
I've tried to test it by using a wrong password or wrong server name. I couldn't figure out how to do a proper test where one server becomes unavailable while it's running. I set the "Will resume..." message as debug because otherwise there would be two lines for every retry. Not sure what is preferable.

### CODE REVIEW

**Comment by jcfp** [comment]
> I couldn't figure out how to do a proper test where one server becomes unavailable while it's running.

Adding a nonexistent server (e.g. some unused port on localhost, with _required_ set) combined with a working server (sabnews test instance at a lower prio) might work? With both the client and server side controlled by the test environment, checking the lower-prio server for activity could be used to verify the _required_ flag.

File: sabnzbd/skintext.py [hunk 1]
```diff
@@ -644,6 +644,10 @@
     "ssl_verify-strict": TT("Strict"),
     "srv-priority": TT("Priority"),  #: Server priority
     "explain-svrprio": TT("0 is highest priority, 100 is the lowest priority"),  #: Explain server priority
+    "srv-required": TT("Required"),  #: Server required tickbox
+    "explain-required": TT(
+        "Download will be paused while connection or login fails for this server"
```
Comment (Safihre) [inline sabnzbd/skintext.py:649]: I don't think this explains the actual way it works:
`In case of connection failures, the download queue will be paused for a few minutes instead of skipping this server`. Or something like that.

File: sabnzbd/scheduler.py [hunk 2]
```diff
@@ -383,6 +383,12 @@ def __check_diskspace(self, full_dir: str, required_space: float):
         if not sabnzbd.Downloader.paused:
             self.cancel_resume_task()
 
+    def __resume_downloading(self):
+        """Unpause Downloader if paused"""
+        if sabnzbd.Downloader.paused:
```
Comment (Safihre) [inline sabnzbd/scheduler.py:388]: Seems the check for it being paused is redundant? The `Downloader.resume()` also performs this check.

File: sabnzbd/scheduler.py [hunk 3]
```diff
@@ -391,11 +397,24 @@ def plan_diskspace_resume(self, full_dir: str, required_space: float):
             self.__check_diskspace, "check_diskspace", 5 * 60, 9 * 60, "threaded", args=[full_dir, required_space]
         )
 
+    def plan_required_server_resume(self, interval: int = 5):
```
Comment (Safihre) [inline sabnzbd/scheduler.py:400]: I wonder if we should use `plan_resume()` in this case? This way the user gets a visual indication how long the pause will last until it tries again.
This is different from the diskspace pause, because for the diskspace pause, it gives a warning before pausing. But here, the queue would just switch to paused mode and magically unpause again after a while.

**Comment by puzzledsab** [comment]
It took a while before I figured out that using the paused flag was the easiest way to avoid doing duplicate checking for each connection. The __resume_downloading stuff was legacy from that which I didn't realize was redundant. I mean to look into what jcfp tipped about but I won't have time for it in the next few days.

File: sabnzbd/downloader.py [hunk 4]
```diff
@@ -835,12 +844,17 @@ def run(self):
...
+                                retry_article = False
                                 if server.active:
-                                    server.deactivate()
-                                    if penalty and (block or server.optional):
-                                        self.plan_server(server, penalty)
-                                # Note that the article is discard for this server
-                                self.__reset_nw(nw, retry_article=False, send_quit=True)
+                                    if server.required:
+                                        sabnzbd.Scheduler.plan_required_server_resume()
+                                        retry_article = True
```
Comment (Safihre) [inline sabnzbd/downloader.py:851]: Good catch!

File: sabnzbd/scheduler.py [hunk 5]
```diff
@@ -391,11 +391,19 @@ def plan_diskspace_resume(self, full_dir: str, required_space: float):
...
+
     def cancel_resume_task(self):
         """Cancel the current auto resume task"""
         if self.resume_task:
             logging.debug("Cancelling existing resume_task '%s'", self.resume_task.name)
-            self.scheduler.cancel(self.resume_task)
+            try:
+                self.scheduler.cancel(self.resume_task)
+            except ValueError:
+                pass
```
Comment (Safihre) [inline sabnzbd/scheduler.py:406]: Is this change still needed now `plan_resume` is used?

**Review by Safihre — APPROVED** [review]

**Comment by Safihre** [comment]
Will you also update the wiki?
I see we also didn't add `process_unpacked_par2` yet.
